### PR TITLE
Bootstrap package requests

### PR DESCRIPTION
**Pull request by bgeuken**

### CODE REVIEW

File: src/api/app/views/webui/shared/_requests_state_and_type_form.html.haml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  = select_tag('request_type_select',
+               options_for_select(@available_types, @default_request_type),
+               data: { table: 'all_requests_table' })
+  requests in state
+  = select_tag('request_state_select',
+               options_for_select(@available_states, @default_request_state),
+               data: { table: 'all_requests_table' })
+  %span#request_count
+  \:
+  = image_tag('ajax-loader.gif', id: 'spinner', class: 'hidden')
```
Comment (hennevogel) [inline src/api/app/views/webui/shared/_requests_state_and_type_form.html.haml]: This is [`.invisible`](https://getbootstrap.com/docs/4.1/utilities/visibility/) in bootstrap...

**Comment by hennevogel** [comment]
I can give it a shot if you want to...

**Comment by bgeuken** [comment]
* Rebased with the other request related changes (js and haml). That previously broke the spinner.
* Added font awesome spinner icons
* Wrapped the tab content in a card. That also fixed the responsiveness issue.

File: src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+%p
+  Display
+  = select_tag('request_type_select',
```
Comment (hennevogel) [inline src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml]: You forgot the `.form-control` class. Also the whole thing is rather crowded and unaligned. 

screen shot 2018-08-22 at 00 45 59

File: src/api/app/views/webui2/shared/_requests_table.html.haml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+%table.requests-datatable.table.table-striped.table-bordered{ 'data-source': source_url,
```
Comment (hennevogel) [inline src/api/app/views/webui2/shared/_requests_table.html.haml]: There are [responsive tables](https://getbootstrap.com/docs/4.1/content/tables/#responsive-tables) in bootstrap, there is also an [datatables extension](https://datatables.net/extensions/responsive/) to do this, with bootstrap integration. Check it out, or get a card in the backlog about this going please.

**Comment by mdeniz** [comment]
Something strange is happening to the table in full size:

image

**Comment by bgeuken** [comment]
I am still working on updating the view

**Comment by mdeniz** [comment]
Hey, whenever you are done with this PR... close it and open another to the `bootstrap-master` branch instead.

**Comment by bgeuken** [comment]
Why bootstrap-master?

**Comment by mdeniz** [comment]
> Why bootstrap-master?

Remember that is in that branch where we are going to add changes for bootstrap instead of master. The point was to have smaller PRs that may break the workflows and until a page/workflow is fully working it shouldn't be merged on master as we deploy that on our instance.

**Comment by bgeuken** [comment]
I consider this PR quite small and it's isolated from the old webui workflow and ui

**Comment by mdeniz** [comment]
> I consider this PR quite small and it's isolated from the old webui workflow and ui

Ok, keep it that way. But, for the next ones we must to follow our own agreements :+1:

**Comment by bgeuken** [comment]
> mdeniz commented 
> I consider this PR quite small and it's isolated from the old webui workflow and uiOk, keep it that way. But, for the next ones we must to follow our own agreements

Well, I talked with Henne before and he confirmed. So I follow our agreements

Or to cite the initial mail regarding this:
`We can only merge complete working pages into master`

And this is the case here:+1:

**Comment by bgeuken** [comment]
@hennevogel @mdeniz @dmarcoux Updated the PR again. Took a bit longer than I hoped. Sorry for that.

One remaining issue:
* The 'Show x entries' and Search bar are getting centered at a certain (small enough) screen size.
  That's because both elements are generated by datatable, while the two select boxes are customized.

I find that a bit annoying, but I also don't want to write lot's custom css for this. Since it's kind of an edge case

File: src/api/app/views/webui2/webui/package/requests.html.haml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+- @pagetitle = "Requests for #{@package}"
+- package_bread_crumb('Requests')
```
[the remark this answers is not in the document]
Comment (dmarcoux) [reply src/api/app/views/webui2/webui/package/requests.html.haml]: The line should be removed now that it is in the breadcrumbs partial.

**Comment by hennevogel** [comment]
> I also don't want to write lot's custom css for this

Well we can create a card for this, no problem.

File: src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+                   data: { table: 'all_requests_table' },
+                   class: 'form-control custom-select custom-select-sm form-control-sm dataTables_length')
+.row
+  .form-group
+    .col-sm-12
+      = label_tag('request_state_select', 'Requests in state', class: 'control-label')
+      = select_tag('request_state_select',
+                   options_for_select(Webui::RequestHelper::AVAILABLE_STATES, selected_state),
+                   data: { table: 'all_requests_table' },
+                   class: 'form-control custom-select custom-select-sm form-control-sm dataTables_length')
```
Comment (hennevogel) [inline src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml]: Now this looks a bit strange on the larger screen. 

screenshot from 2018-08-22 17-13-20

How about you move the selects into partials and then you can render them more than one time and [hide](https://getbootstrap.com/docs/4.1/utilities/display/#hiding-elements) what you don't need in the different breakpoints.

Comment (bgeuken) [reply src/api/app/views/webui2/shared/_requests_state_and_type_form.html.haml]: Updated. I found the CSS for setting correct breakpoints for different screensizes. So no need for hiding and showing elements:tada:

**Comment by bgeuken** [comment]
If you know of a better icon for the request links let me know:wink:

File: src/api/app/controllers/webui/packages/bs_requests_controller.rb [hunk 6]
```diff
@@ -13,7 +13,11 @@ def index
         session[:request_numbers] = requests_query.requests.map(&:number)
 
         respond_to do |format|
-          format.json
+          if switch_to_webui2?
+            format.json { render 'webui2/shared/bs_requests/index' }
+          else
+            format.json
+          end
```
Comment (hennevogel) [inline src/api/app/controllers/webui/packages/bs_requests_controller.rb]: Interesting way of doing this :)

File: src/api/spec/support/capybara.rb [hunk 7]
```diff
@@ -12,6 +12,7 @@
     browser_options = ::Selenium::WebDriver::Chrome::Options.new
     browser_options.args << '--headless'
     browser_options.args << '--no-sandbox' # to run in docker
+    browser_options.args << '--window-size=1280,1024'
```
Comment (hennevogel) [inline src/api/spec/support/capybara.rb]: What was it before?

File: src/api/app/views/webui2/shared/bs_requests/index.json.erb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    <% @requests_data_table.rows.each.with_index do |row,index | %>
+      [
+        "<%= escape_javascript(fuzzy_time(row.created_at, false)) %>",
+        <% cache "request-table-row-#{row.id}-#{row.updated_at.to_i}" do %>
+          "<%= escape_javascript(project_or_package_link(project: row.source_project, package: row.source_package, creator: row.creator, trim_to: 40, short: true)) %>",
+          "<%= escape_javascript(target_project_link(row)) %>",
+          "<%= escape_javascript(user_with_realname_and_icon(row.creator, short: true)) %>",
+          "<%= escape_javascript(new_or_update_request(row)) %>",
+          "<%= escape_javascript(row.priority) %>",
+          "<%= escape_javascript(link_to(icon('fa', 'info-circle'), request_show_path(row.number), { title: "Show request ##{row.number}", class: :request_link })) %>"
```
Comment (hennevogel) [inline src/api/app/views/webui2/shared/bs_requests/index.json.erb]: Oh please avoid the bloody helpers if you can. I would really appreciate it if we could get rid of all this nonsense and pull only over what we really need...

But we can do this later, I'm fine with a FIXME...

File: src/api/spec/support/capybara.rb (hunk 7 above)
Comment (bgeuken) [reply src/api/spec/support/capybara.rb]: I couldn't find that in the documentation, but it looked like ~500px

File: src/api/app/views/webui2/shared/bs_requests/index.json.erb (hunk 8 above)
Comment (bgeuken) [reply src/api/app/views/webui2/shared/bs_requests/index.json.erb]: The `icon` call? What's so bad about it?

File: src/api/app/controllers/webui/packages/bs_requests_controller.rb (hunk 6 above)
Comment (bgeuken) [reply src/api/app/controllers/webui/packages/bs_requests_controller.rb]: The `switch_to_webui`  doesn't seem to work when using a respond_to block:/

File: src/api/app/views/webui2/webui/package/requests.html.haml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bgeuken) [reply src/api/app/views/webui2/webui/package/requests.html.haml]: Oh, thanks. I am blind:blush:

**Comment by dmarcoux** [comment]
Looking at a page with no requests, I find a few elements a bit strange.

- The text `Showing 0 to 0 of 0 entries`. That's a lot of zeros... I would hide this when there are no entries.
- The pagination links `First`, `Next` (etc...) also look out of place in this context. I'm not sure if hiding them is the way to go here. Just wanted to state that I find this weird too.
- The last column in the datatable after `Priority` does not have a name and it looks like something is wrong with the datatable when there is no data. Maybe add some text like `Actions`

no_data

**Comment by bgeuken** [comment]
I think that's fine. It's the default datatables behaviour and the pagination buttons are disabled when there is no previous / next request page.
Also I would find it a bit distracting when these elements appear and disappear all the time.

To me a  'Showing 0 of 0 entries' would still be useful since it tells me that there are no results for my request state / type selection.

**Comment by bgeuken** [comment]
Regarding the priority sorting. That works (just tested on my machine). In the review app it does not because both requests have the same priority.

And for the action label... I would be fine with that. But I would prefer to do that in a follow up PR. Right now this is like it was in the old UI and changing that might mess up the column size.

File: src/api/app/views/webui2/shared/bs_requests/index.json.erb (hunk 8 above)
Comment (bgeuken) [reply src/api/app/views/webui2/shared/bs_requests/index.json.erb]: I see, though [icon](https://github.com/FortAwesome/font-awesome-sass#rails-helper-usage) is a at least not a custom one.

**Pull request merged**